### PR TITLE
Deduplicate fs error handling

### PR DESCRIPTION
**Pull request by jviotti**
See: https://github.com/resin-io-modules/etcher-image-write/pull/96#discussion_r112429753
Signed-off-by: Juan Cruz Viotti <jviotti@openmailbox.org>

### CODE REVIEW

**Comment by lurch** [comment]
Requested a review from Jonas, as this is advanced NodeJS code I can't mentally parse yet ;-)

**Comment by jviotti** [comment]
You might want to install Esprima (http://esprima.org/) on yourself :P

**Comment by lurch** [comment]
`eslint` is making the CI tests fail.

I tried installing Esprima on myself, but my brain's JS interpreter was too old to handle it :-(

File: lib/filesystem.js [hunk 1]
```diff
@@ -40,6 +40,63 @@ const MAXIMUM_RETRIES = 10;
...
 /**
+ * @summary Handle read/write errors
+ * @function
+ * @private
+ *
+ * @param {Error} error - error
+ * @param {Object} options - options
+ * @param {Number} options.currentRetries - current number of retries
+ * @param {Function} options.retry - retry function (currentRetries)
+ * @param {Function} callback - callback (error)
```
Comment (jhermsmeier) [inline lib/filesystem.js]: A missing `@returns undefined` is what the linter is complaining about here

File: lib/filesystem.js [hunk 2]
```diff
@@ -40,6 +40,63 @@ const MAXIMUM_RETRIES = 10;
...
+
+      return callback(error);
+    }
+
+    return setTimeout(() => {
+      options.retry(options.currentRetries);
+    }, RETRY_BASE_TIMEOUT * options.currentRetries);
+  }
+
+  return callback(error);
```
Comment (jhermsmeier) [inline lib/filesystem.js]: No need to `return` here, I believe

File: lib/filesystem.js [hunk 3]
```diff
@@ -40,6 +40,63 @@ const MAXIMUM_RETRIES = 10;
...
+
+  // In some faulty SDCards or SDCard readers you might randomly
+  // get EIO errors, but they usually go away after some retries.
+  if (error.code === 'EIO') {
+    if (options.currentRetries > MAXIMUM_RETRIES) {
+      if (platform === 'linux') {
+        return callback(errors.unpluggedError());
+      }
+
+      return callback(error);
```
Comment (jhermsmeier) [inline lib/filesystem.js]: Just a little nit; If you want to ensure the same return value everywhere – in this case `undefined` – you can do `return void callback(error)`, as `void` will make any expression return `undefined`

File: lib/filesystem.js (hunk 1 above)
Comment (jviotti) [inline lib/filesystem.js]: Fixed!

File: lib/filesystem.js (hunk 2 above)
Comment (jviotti) [inline lib/filesystem.js]: Fixed!

File: lib/filesystem.js (hunk 3 above)
Comment (jviotti) [inline lib/filesystem.js]: Cool stuff. I don't think we need to make sure we return undefined here though.

Comment (jhermsmeier) [inline lib/filesystem.js]: Yeah, was just thinking about that for ultimate correctness of the docs, as the callbacks might actually returns something else, but shouldn't really matter

**Pull request merged**